lockc/src/uprobe_ext.rs

```diff
@@ -0,0 +1,163 @@
+        })
+    }
+
+    /// Find the offset of a symbol in the ELF file
+    pub fn find_offset(&self, symbol: &str) -> Option<usize> {
+        self.resolve_sym(symbol).map(|sym| sym.st_value as usize)
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum AttachUprobeSymbolError {
+    #[error(transparent)]
+    Libbpf(#[from] libbpf_rs::Error),
+
+    #[error(transparent)]
+    FindInFile(#[from] FindInFileError),
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum AttachUprobeAddrError {
+    #[error(transparent)]
+    Libbpf(#[from] libbpf_rs::Error),
+
+    #[error(transparent)]
+    Proc(#[from] procfs::ProcError),
+
+    #[error("failed to find executable region")]
+    NotFound,
+}
+
+pub trait FindSymbolUprobeExt {
+    fn attach_uprobe_symbol(
+        &mut self,
+        retprobe: bool,
+        pid: i32,
+        pathname: &Path,
+        symbol: &str,
+    ) -> Result<libbpf_rs::Link, AttachUprobeSymbolError>;
+
+    fn attach_uprobe_addr(
+        &mut self,
+        retprobe: bool,
+        pid: i32,
+        addr: usize,
+    ) -> Result<libbpf_rs::Link, AttachUprobeAddrError>;
+}
+
+impl FindSymbolUprobeExt for libbpf_rs::Program {
+    /// Attach a uprobe to a symbol within another binary.
+    fn attach_uprobe_symbol(
+        &mut self,
+        retprobe: bool,
+        pid: i32,
+        pathname: &Path,
+        symbol: &str,
+    ) -> Result<libbpf_rs::Link, AttachUprobeSymbolError> {
+        // Find symbol in the ELF file
+        let offset = SymbolResolver::find_in_file(pathname, symbol)?;
+
+        // Use the offset we found to attach the probe
+        match self.attach_uprobe(retprobe, pid, pathname, offset) {
+            Ok(link) => Ok(link),
+            Err(e) => Err(AttachUprobeSymbolError::from(e)),
+        }
+    }
+
+    /// Attach a uprobe to an address within our own address space.
+    fn attach_uprobe_addr(
+        &mut self,
+        retprobe: bool,
+        pid: i32,
+        addr: usize,
+    ) -> Result<libbpf_rs::Link, AttachUprobeAddrError> {
+        // Find the offset
+        let base_addr = get_base_addr()?;
+        let offset = addr - base_addr;
+
+        let pathname = "/proc/self/exe";
+
+        // Use the offset we found to attach the probe
+        match self.attach_uprobe(retprobe, pid, pathname, offset) {
+            Ok(link) => Ok(link),
+            Err(e) => Err(AttachUprobeAddrError::from(e)),
+        }
+    }
+}
+
+/// Find our base load address. We use /proc/self/maps for this.
+fn get_base_addr() -> Result<usize, AttachUprobeAddrError> {
+    let me = Process::myself()?;
+    let maps = me.maps()?;
+
+    for entry in maps {
+        if entry.perms.contains("r-xp") {
+            return Ok((entry.address.0 - entry.offset) as usize);
+        }
+    }
+
+    Err(AttachUprobeAddrError::NotFound)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn get_base_addr_smoke_test() {
+        get_base_addr().expect("Calling get_base_addr failed");
+    }
+}
```

Comment:
May I suggest that we also include a test on a real ELF binary? It doesn't have to be anything complex. Again, this could probably wait until we upstream this in Aya.

Reply:
Yes, I would prefer to do that later with Aya. I don't want to put too much effort into libbpf-rs based code.